### PR TITLE
external authz: Interpret error grpc codes of the external authz server as failure to fix failure_mode_allowed feature

### LINKED ISSUE #34705: `failure_mode_allow=true` is not fully supported for the grpc authz server.
*Title*: `failure_mode_allow=true` is not fully supported for the grpc authz server.

*Description*:
There is inconsistency in how the `failure_mode_allow=true` option is applied to external authorization servers depending on the protocol: http and grpc.

In case of http authz server, any 5xx statuses lead to accepting the incoming request (fail-open).

In case of grpc authz server, statuses like 14 (Unavailable) lead to rejecting the incoming request. If there is any grpc proxy between envoy and authz server, then this proxy can return 14 (Unavailable) when authz server is unavailable. It leads to the fail-close behaviour even when `failure_mode_allow` is enabled. As a result, we have random rejections while with `failure_mode_allow` we want the opposite - don't reject requests because of random failures.
 
Expected behaviour is to use failue_mode_allow when receiveng grpc statuses like 2 (Unknown), 13 (Internal), 14 (Unavailable).

The following PR https://github.com/envoyproxy/envoy/pull/4199 was intended to handle these cases when `failure_mode_allow=true` and authz server is unhealthy, but it was solved for HTTP authz server only.

*Repro steps*:
1. Configure envoy to have `failure_mode_allow=true`
2. Make the authz server to return grpc status 14 (Unavailable)
3. Perform http request to a service in the cluster
4. Envoy returns 403 (fail-close), while it was expected to perform fail-open.

### PR DESCRIPTION
**Pull request by konstantin-baidin-y42**
Commit Message:  Interpret error grpc codes of the external authz server as failure to fix failure_mode_allowed feature
Additional Description: Fixes issue https://github.com/envoyproxy/envoy/issues/34705. Mark authz response as denied only for grpc codes `PermissionDenied` and `Unauthenticated`. Other codes are considered as failure and authz response status is set to error. This new behaviour is aligned with the HTTP external authz checks implementation, where error HTTP codes are considered as failure too. It shouldn't change anything when `failure_mode_allowed=false` as any non-OK grpc codes will lead to 403 for a client. But it fixes behaviour of `failure_mode_allowed=true`. 
Risk Level: Low
Testing: unit test
Docs Changes: ---
Release Notes: With failure_mode_allowed enabled and external grpc authz server returning error codes like internal error, the request will be allowed (fail-open), see https://github.com/envoyproxy/envoy/issues/34705 for details.
Platform Specific Features: ---
Optional Runtime guard: I'm not sure if it should be guarded. It affects end users, but it's a bug fix.
Fixes #34705

### CODE REVIEW

**Comment by htuch** [comment]
Probably needs at least runtime guarding given potential for differences in dataplane behavior. @tyxia should the semantics here be harmonized with ext_proc, in particular if we start using it for auth purposes (CC @rshriram)?

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc [hunk 1]
```diff
@@ -114,7 +114,9 @@ void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::service::auth::v3::CheckRe
       const auto& ok_response = response->ok_response();
       copyOkResponseMutations(authz_response, ok_response);
     }
-  } else {
+  } else if (response->status().code() == Grpc::Status::WellKnownGrpcStatus::PermissionDenied ||
```
Comment (ggreenway) [inline source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:117]: Please add a runtime guard for this behavior change.

**Review by ggreenway — CHANGES_REQUESTED** [review]
Thanks for working on this!

This change needs a runtime guard around the behavior change, and a changelog entry.

/wait

File: source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc (hunk 1 above)
Comment (konstantin-baidin-y42) [reply source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc:117]: I added a runtime guard and a changelog entry. Can you please take a look?

File: test/extensions/filters/common/ext_authz/ext_authz_grpc_impl_test.cc [hunk 2]
```diff
@@ -436,6 +437,46 @@ TEST_F(ExtAuthzGrpcClientTest, AuthorizationOkWithAppendActions) {
   client_->onSuccess(std::make_unique<envoy::service::auth::v3::CheckResponse>(check_response),
                      span_);
 }
+// Test the client when an error code response is received.
+TEST_F(ExtAuthzGrpcClientTest, AuthorizationReturnsErrorOnBadGrpcCode) {
+  TestScopedRuntime scoped_runtime;
+  scoped_runtime.mergeValues(
+      {{"envoy.reloadable_features.process_ext_authz_grpc_error_codes_as_errors", "true"}});
```
Comment (ggreenway) [inline test/extensions/filters/common/ext_authz/ext_authz_grpc_impl_test.cc:444]: Please add a second test case that verifies this fails when the runtime guard is set to false.

File: changelogs/current.yaml [hunk 3]
```diff
@@ -246,6 +246,10 @@ bug_fixes:
 - area: datadog
   change: |
     Bumped the version of datadog to resolve a crashing bug in earlier versions of the library.
+- area: ext_authz
+  change: |
+    Fixed fail-open behaviour of the failure_mode_allow config option when a grpc external authz server is used.
```
Comment (ggreenway) [inline changelogs/current.yaml:251]: Please add a link to the `failure_mode_allow` field in the docs

**Review by ggreenway — CHANGES_REQUESTED** [review]
Please clarify in the docs for `failure_mode_allow` (in ext_authz.proto) which grpc response codes are still treated as `denied` when this is true.

/wait

File: changelogs/current.yaml (hunk 3 above)
Comment (konstantin-baidin-y42) [reply changelogs/current.yaml]: Looks like the https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto page doesn't have anchor links, so I added just a link to the page where this param can be found

**Comment by konstantin-baidin-y42** [comment]
Hello @ggreenway, thank you for the review.

- I clarified in the docs for failure_mode_allow (in ext_authz.proto) which grpc response codes are still treated as denied when this is true. 
- The runtime guard is now true by default.
- I replaced the old test `AuthorizationDeniedGrpcUnknownStatus` with the new test `AuthorizationReturnsDeniedOnUnknownGrpcCodeWhenFeatureFlagDisabled`
- I tried to add a link for failure_mode_allow docs, but it looks like the https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto page doesn't have anchor links, so I added just a link to the page where this param can be found.

Can you please review again?

**Review by htuch — COMMENTED** [review]
/lgtm api

File: changelogs/current.yaml [hunk 4]
```diff
@@ -267,6 +267,12 @@ bug_fixes:
 - area: lua
   change: |
     Fixed a bug where the user data will reference a dangling pointer to the Lua state and cause a crash.
+- area: ext_authz
+  change: |
+    Fixed fail-open behaviour of the failure_mode_allow config option
+    https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto
```
Comment (ggreenway) [inline changelogs/current.yaml:273]: Sorry I wasn't clear about the link. It should be something like 
```suggestion
    Fixed fail-open behaviour of the :ref:`failure_mode_allow config option 
    <envoy_v3_api_field_extensions.filters.http.ext_authz.v3.ExtAuthz.failure_mode_allow>`
```

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

**Review by ggreenway — DISMISSED** [review]
LGTM, but has a conflict on the changelog. Please resolve, then it's ready to merge.

/wait

**Comment by konstantin-baidin-y42** [comment]
Resolved conflict

**Review by ggreenway — APPROVED** [review]

**Pull request merged**